<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="application/xhtml+xml; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.6.9" />
<title>RUNNING IGGPIPE TO GENERATE IGG MARKERS</title>
<style type="text/css">
/* Shared CSS for AsciiDoc xhtml11 and html5 backends */

/* Default font. */
body {
  font-family: Georgia,serif;
}

/* Title font. */
h1, h2, h3, h4, h5, h6,
div.title, caption.title,
thead, p.table.header,
#toctitle,
#author, #revnumber, #revdate, #revremark,
#footer {
  font-family: Arial,Helvetica,sans-serif;
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}
h5 {
  font-size: 1.0em;
}

div.sectionbody {
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}
ul > li     { color: #aaa; }
ul > li > * { color: black; }

.monospaced, code, pre {
  font-family: "Courier New", Courier, monospace;
  font-size: inherit;
  color: navy;
  padding: 0;
  margin: 0;
}
pre {
  white-space: pre-wrap;
}

#author {
  color: #527bbd;
  font-weight: bold;
  font-size: 1.1em;
}
#email {
}
#revnumber, #revdate, #revremark {
}

#footer {
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

#preamble {
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.0em;
  margin-bottom: 2.0em;
  margin-right: 10%;
  color: #606060;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid #dddddd;
  border-left: 4px solid #f0f0f0;
  padding: 0.5em;
}

div.listingblock > div.content {
  border: 1px solid #dddddd;
  border-left: 5px solid #f0f0f0;
  background: #f8f8f8;
  padding: 0.5em;
}

div.quoteblock, div.verseblock {
  padding-left: 1.0em;
  margin-left: 1.0em;
  margin-right: 10%;
  border-left: 5px solid #f0f0f0;
  color: #888;
}

div.quoteblock > div.attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock > pre.content {
  font-family: inherit;
  font-size: inherit;
}
div.verseblock > div.attribution {
  padding-top: 0.75em;
  text-align: left;
}
/* DEPRECATED: Pre version 8.2.7 verse style literal block. */
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 3px solid #dddddd;
}

div.exampleblock > div.content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

div.imageblock div.content { padding-left: 0; }
span.image img { border-style: none; vertical-align: text-bottom; }
a.image:visited { color: white; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
  color: navy;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
ol.arabic {
  list-style-type: decimal;
}
ol.loweralpha {
  list-style-type: lower-alpha;
}
ol.upperalpha {
  list-style-type: upper-alpha;
}
ol.lowerroman {
  list-style-type: lower-roman;
}
ol.upperroman {
  list-style-type: upper-roman;
}

div.compact ul, div.compact ol,
div.compact p, div.compact p,
div.compact div, div.compact div {
  margin-top: 0.1em;
  margin-bottom: 0.1em;
}

tfoot {
  font-weight: bold;
}
td > div.verse {
  white-space: pre;
}

div.hdlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hdlist tr {
  padding-bottom: 15px;
}
dt.hdlist1.strong, td.hdlist1.strong {
  font-weight: bold;
}
td.hdlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
  color: navy;
}
td.hdlist2 {
  vertical-align: top;
}
div.hdlist.compact tr {
  margin: 0;
  padding-bottom: 0;
}

.comment {
  background: yellow;
}

.footnote, .footnoteref {
  font-size: 0.8em;
}

span.footnote, span.footnoteref {
  vertical-align: super;
}

#footnotes {
  margin: 20px 0 20px 0;
  padding: 7px 0 0 0;
}

#footnotes div.footnote {
  margin: 0 0 5px 0;
}

#footnotes hr {
  border: none;
  border-top: 1px solid silver;
  height: 1px;
  text-align: left;
  margin-left: 0;
  width: 20%;
  min-width: 100px;
}

div.colist td {
  padding-right: 0.5em;
  padding-bottom: 0.3em;
  vertical-align: top;
}
div.colist td img {
  margin-top: 0.3em;
}

@media print {
  #footer-badges { display: none; }
}

#toc {
  margin-bottom: 2.5em;
}

#toctitle {
  color: #527bbd;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel0, div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}

span.aqua { color: aqua; }
span.black { color: black; }
span.blue { color: blue; }
span.fuchsia { color: fuchsia; }
span.gray { color: gray; }
span.green { color: green; }
span.lime { color: lime; }
span.maroon { color: maroon; }
span.navy { color: navy; }
span.olive { color: olive; }
span.purple { color: purple; }
span.red { color: red; }
span.silver { color: silver; }
span.teal { color: teal; }
span.white { color: white; }
span.yellow { color: yellow; }

span.aqua-background { background: aqua; }
span.black-background { background: black; }
span.blue-background { background: blue; }
span.fuchsia-background { background: fuchsia; }
span.gray-background { background: gray; }
span.green-background { background: green; }
span.lime-background { background: lime; }
span.maroon-background { background: maroon; }
span.navy-background { background: navy; }
span.olive-background { background: olive; }
span.purple-background { background: purple; }
span.red-background { background: red; }
span.silver-background { background: silver; }
span.teal-background { background: teal; }
span.white-background { background: white; }
span.yellow-background { background: yellow; }

span.big { font-size: 2em; }
span.small { font-size: 0.6em; }

span.underline { text-decoration: underline; }
span.overline { text-decoration: overline; }
span.line-through { text-decoration: line-through; }

div.unbreakable { page-break-inside: avoid; }


/*
 * xhtml11 specific
 *
 * */

div.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.tableblock > table {
  border: 3px solid #527bbd;
}
thead, p.table.header {
  font-weight: bold;
  color: #527bbd;
}
p.table {
  margin-top: 0;
}
/* Because the table frame attribute is overriden by CSS in most browsers. */
div.tableblock > table[frame="void"] {
  border-style: none;
}
div.tableblock > table[frame="hsides"] {
  border-left-style: none;
  border-right-style: none;
}
div.tableblock > table[frame="vsides"] {
  border-top-style: none;
  border-bottom-style: none;
}


/*
 * html5 specific
 *
 * */

table.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
thead, p.tableblock.header {
  font-weight: bold;
  color: #527bbd;
}
p.tableblock {
  margin-top: 0;
}
table.tableblock {
  border-width: 3px;
  border-spacing: 0px;
  border-style: solid;
  border-color: #527bbd;
  border-collapse: collapse;
}
th.tableblock, td.tableblock {
  border-width: 1px;
  padding: 4px;
  border-style: solid;
  border-color: #527bbd;
}

table.tableblock.frame-topbot {
  border-left-style: hidden;
  border-right-style: hidden;
}
table.tableblock.frame-sides {
  border-top-style: hidden;
  border-bottom-style: hidden;
}
table.tableblock.frame-none {
  border-style: hidden;
}

th.tableblock.halign-left, td.tableblock.halign-left {
  text-align: left;
}
th.tableblock.halign-center, td.tableblock.halign-center {
  text-align: center;
}
th.tableblock.halign-right, td.tableblock.halign-right {
  text-align: right;
}

th.tableblock.valign-top, td.tableblock.valign-top {
  vertical-align: top;
}
th.tableblock.valign-middle, td.tableblock.valign-middle {
  vertical-align: middle;
}
th.tableblock.valign-bottom, td.tableblock.valign-bottom {
  vertical-align: bottom;
}


/*
 * manpage specific
 *
 * */

body.manpage h1 {
  padding-top: 0.5em;
  padding-bottom: 0.5em;
  border-top: 2px solid silver;
  border-bottom: 2px solid silver;
}
body.manpage h2 {
  border-style: none;
}
body.manpage div.sectionbody {
  margin-left: 3em;
}

@media print {
  body.manpage div#toc { display: none; }
}


</style>
<script type="text/javascript">
/*<![CDATA[*/
var asciidoc = {  // Namespace.

/////////////////////////////////////////////////////////////////////
// Table Of Contents generator
/////////////////////////////////////////////////////////////////////

/* Author: Mihai Bazon, September 2002
 * http://students.infoiasi.ro/~mishoo
 *
 * Table Of Content generator
 * Version: 0.4
 *
 * Feel free to use this script under the terms of the GNU General Public
 * License, as long as you do not remove or alter this notice.
 */

 /* modified by Troy D. Hanson, September 2006. License: GPL */
 /* modified by Stuart Rackham, 2006, 2009. License: GPL */

// toclevels = 1..4.
toc: function (toclevels) {

  function getText(el) {
    var text = "";
    for (var i = el.firstChild; i != null; i = i.nextSibling) {
      if (i.nodeType == 3 /* Node.TEXT_NODE */) // IE doesn't speak constants.
        text += i.data;
      else if (i.firstChild != null)
        text += getText(i);
    }
    return text;
  }

  function TocEntry(el, text, toclevel) {
    this.element = el;
    this.text = text;
    this.toclevel = toclevel;
  }

  function tocEntries(el, toclevels) {
    var result = new Array;
    var re = new RegExp('[hH]([1-'+(toclevels+1)+'])');
    // Function that scans the DOM tree for header elements (the DOM2
    // nodeIterator API would be a better technique but not supported by all
    // browsers).
    var iterate = function (el) {
      for (var i = el.firstChild; i != null; i = i.nextSibling) {
        if (i.nodeType == 1 /* Node.ELEMENT_NODE */) {
          var mo = re.exec(i.tagName);
          if (mo && (i.getAttribute("class") || i.getAttribute("className")) != "float") {
            result[result.length] = new TocEntry(i, getText(i), mo[1]-1);
          }
          iterate(i);
        }
      }
    }
    iterate(el);
    return result;
  }

  var toc = document.getElementById("toc");
  if (!toc) {
    return;
  }

  // Delete existing TOC entries in case we're reloading the TOC.
  var tocEntriesToRemove = [];
  var i;
  for (i = 0; i < toc.childNodes.length; i++) {
    var entry = toc.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div'
     && entry.getAttribute("class")
     && entry.getAttribute("class").match(/^toclevel/))
      tocEntriesToRemove.push(entry);
  }
  for (i = 0; i < tocEntriesToRemove.length; i++) {
    toc.removeChild(tocEntriesToRemove[i]);
  }

  // Rebuild TOC entries.
  var entries = tocEntries(document.getElementById("content"), toclevels);
  for (var i = 0; i < entries.length; ++i) {
    var entry = entries[i];
    if (entry.element.id == "")
      entry.element.id = "_toc_" + i;
    var a = document.createElement("a");
    a.href = "#" + entry.element.id;
    a.appendChild(document.createTextNode(entry.text));
    var div = document.createElement("div");
    div.appendChild(a);
    div.className = "toclevel" + entry.toclevel;
    toc.appendChild(div);
  }
  if (entries.length == 0)
    toc.parentNode.removeChild(toc);
},


/////////////////////////////////////////////////////////////////////
// Footnotes generator
/////////////////////////////////////////////////////////////////////

/* Based on footnote generation code from:
 * http://www.brandspankingnew.net/archive/2005/07/format_footnote.html
 */

footnotes: function () {
  // Delete existing footnote entries in case we're reloading the footnodes.
  var i;
  var noteholder = document.getElementById("footnotes");
  if (!noteholder) {
    return;
  }
  var entriesToRemove = [];
  for (i = 0; i < noteholder.childNodes.length; i++) {
    var entry = noteholder.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div' && entry.getAttribute("class") == "footnote")
      entriesToRemove.push(entry);
  }
  for (i = 0; i < entriesToRemove.length; i++) {
    noteholder.removeChild(entriesToRemove[i]);
  }

  // Rebuild footnote entries.
  var cont = document.getElementById("content");
  var spans = cont.getElementsByTagName("span");
  var refs = {};
  var n = 0;
  for (i=0; i<spans.length; i++) {
    if (spans[i].className == "footnote") {
      n++;
      var note = spans[i].getAttribute("data-note");
      if (!note) {
        // Use [\s\S] in place of . so multi-line matches work.
        // Because JavaScript has no s (dotall) regex flag.
        note = spans[i].innerHTML.match(/\s*\[([\s\S]*)]\s*/)[1];
        spans[i].innerHTML =
          "[<a id='_footnoteref_" + n + "' href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
        spans[i].setAttribute("data-note", note);
      }
      noteholder.innerHTML +=
        "<div class='footnote' id='_footnote_" + n + "'>" +
        "<a href='#_footnoteref_" + n + "' title='Return to text'>" +
        n + "</a>. " + note + "</div>";
      var id =spans[i].getAttribute("id");
      if (id != null) refs["#"+id] = n;
    }
  }
  if (n == 0)
    noteholder.parentNode.removeChild(noteholder);
  else {
    // Process footnoterefs.
    for (i=0; i<spans.length; i++) {
      if (spans[i].className == "footnoteref") {
        var href = spans[i].getElementsByTagName("a")[0].getAttribute("href");
        href = href.match(/#.*/)[0];  // Because IE return full URL.
        n = refs[href];
        spans[i].innerHTML =
          "[<a href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
      }
    }
  }
},

install: function(toclevels) {
  var timerId;

  function reinstall() {
    asciidoc.footnotes();
    if (toclevels) {
      asciidoc.toc(toclevels);
    }
  }

  function reinstallAndRemoveTimer() {
    clearInterval(timerId);
    reinstall();
  }

  timerId = setInterval(reinstall, 500);
  if (document.addEventListener)
    document.addEventListener("DOMContentLoaded", reinstallAndRemoveTimer, false);
  else
    window.onload = reinstallAndRemoveTimer;
}

}
asciidoc.install();
/*]]>*/
</script>
</head>
<body class="article">
<div id="header">
<h1>RUNNING IGGPIPE TO GENERATE IGG MARKERS</h1>
<span id="author">Version v0.0, Aug 2015</span><br />
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph"><p>This describes how to run the IGGPIPE software to analyze two or more genome
sequences and produce a file of candidate IGG marker primer pairs for amplifying
length-polymorphic regions of those genomes.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_strong_if_not_done_already_download_iggpipe_files_strong"><strong>If not done already, download IGGPIPE files</strong></h2>
<div class="sectionbody">
<div class="olist arabic"><ol class="arabic">
<li>
<p>
Browse to <a href="https://github.com/BradyLab/IGGPIPE">https://github.com/BradyLab/IGGPIPE</a>
</p>
</li>
<li>
<p>
At bottom of right column on screen, click "Download ZIP" and choose a place to
put it on your computer.
</p>
</li>
<li>
<p>
Unzip the zip file on your computer.
</p>
</li>
<li>
<p>
Rename the unzipped folder from "IGGPIPE-master" to just "IGGPIPE".
</p>
</li>
</ol></div>
</div>
</div>
<div class="sect1">
<h2 id="_strong_if_not_done_already_install_iggpipe_strong"><strong>If not done already, install IGGPIPE</strong></h2>
<div class="sectionbody">
<div class="ulist"><ul>
<li>
<p>
Look inside the downloaded IGGPIPE folder on your computer for file INSTALL.pdf
or INSTALL.html and open either one and follow the instructions.
</p>
</li>
</ul></div>
</div>
</div>
<div class="sect1">
<h2 id="_strong_requirements_for_running_iggpipe_strong"><strong>Requirements for running IGGPIPE</strong></h2>
<div class="sectionbody">
<div class="olist arabic"><ol class="arabic">
<li>
<p>
<strong>Obtain genome FASTA files</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>IGGPIPE uses as input two or more FASTA files of complete genome sequences (at
least one of which is assembled into chromosomes, the other(s) can be in the
scaffold state). You should know ahead of time <strong>which</strong> genomes you are
comparing.  If you haven&#8217;t already, download their genomic FASTA files, which
are required.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Work from the command line</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>Most work here is done from the command line, by opening the Terminal application.
Commands will be shown here and you may be able to get by with no knowledge of the
command line, other than knowing how to start it (by starting the Terminal app
on the Mac). However, ideally you should be familiar with some basic command
line commands such as <em>cd</em>, <em>ls</em>, <em>cp</em>, <em>rm</em>, <em>mv</em>, <em>sudo</em>, <em>less</em>,
<em>head</em> and <em>tail</em>, <em>man</em>, and <em>mkdir</em>. If you do not know how to use the command
line or don&#8217;t know these basic commands, you may want to take some time to learn
a bit. Here are two short tutorial web sites:</p></div>
<div class="paragraph"><p><a href="http://www.davidbaumgold.com/tutorials/command-line">http://www.davidbaumgold.com/tutorials/command-line</a></p></div>
<div class="paragraph"><p><a href="http://mac.appstorm.net/how-to/utilities-how-to/how-to-use-terminal-the-basics">http://mac.appstorm.net/how-to/utilities-how-to/how-to-use-terminal-the-basics</a></p></div>
<div class="paragraph"><p>Here is a longer tutorial from UC Davis:</p></div>
<div class="paragraph"><p><a href="http://korflab.ucdavis.edu/Unix_and_Perl/current.html">http://korflab.ucdavis.edu/Unix_and_Perl/current.html</a></p></div>
</div></div>
</li>
<li>
<p>
<strong>Work in the IGGPIPE main directory unless otherwise instructed</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>While working from the command line to install IGGPIPE, most of the time you will
be in the IGGPIPE main directory, unless instructed otherwise. If you unzipped
the IGGPIPE zip file in your Documents folder, you would change into the IGGPIPE
directory with this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>cd ~/Documents/IGGPIPE</code></pre>
</div></div>
</div></div>
</li>
<li>
<p>
<strong>Know how to use a plain text editor and have one available</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>You must have a plain text editor you know how to use.  If nothing else, the Mac
"TextEdit" program will work (use Plain Text format).  The free open-source
TextWrangler program is strongly recommended, available from the Apple App
Store (Applications, App Store) or from:</p></div>
<div class="paragraph"><p><a href="http://www.barebones.com/products/textwrangler">http://www.barebones.com/products/textwrangler</a></p></div>
</div></div>
</li>
</ol></div>
</div>
</div>
<div class="sect1">
<h2 id="_strong_running_iggpipe_strong"><strong>Running IGGPIPE</strong></h2>
<div class="sectionbody">
<div class="olist arabic"><ol class="arabic">
<li>
<p>
<strong>Assign a single capital letter name to each genome</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>IGGPIPE makes frequent use of a single capital letter to refer to a genome. For
example, filenames and tab-separated file column names use such letters. Choose
a single capital letter you will use to represent each of your genomes. For
example, I used H=Heinz and P=pennellii for some of my testing.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Copy allParameters.template file to new file allParameters.XY</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>IGGPIPE has a number of parameters that must be set to the values you desire. These
are contained in plain text files whose name starts with "allParameters". IGGPIPE
provides a template file named <strong>allParameters.template</strong> that you copy and edit to
create your own "allParameters" files containing the parameter settings you want.</p></div>
<div class="paragraph"><p>You should make a new parameter file for each different set of settings you
want to run with IGGPIPE. For example, each new set of genomes would have a different
parameter file. Also, if you decided to make two different sets of markers from the
same genomes, using two different parameter settings, you might make two different
parameter files. Here, we assume to begin with that you are only running your genomes
with two genomes and a single set of parameters, and we will name the parameter file
allParameters.XY, where X and Y are capital letters you chose to represent the two
genomes. It is convenient to include those letters in the file name, to help you
keep multiple parameter file straight.</p></div>
<div class="paragraph"><p><strong>Copy allParameters.template to allParameters.XY</strong> (substituting your genome letters
for X and Y), using either the Mac&#8217;s Finder or via the command line. For example:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>cd ~/Documents/IGGPIPE    (or whatever is appropriate for your system)
cp allParameters.template allParameters.HP    (here, X=H, Y=P)</code></pre>
</div></div>
<div class="paragraph"><p>Besides the sample parameter file "allParameters.template", there are several more
sample "allParameters" parameter files in the subfolder <em>allParameters</em>.  These
were used for testing IGGPIPE with various genomes.  The subfolder <em>allParameters</em>
was made to keep the files better organized and keep the root folder less messy.
You may want to adopt the same strategy and put your "allParameters" files in the
<em>allParameters</em> subfolder or in a subfolder you make yourself.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Open the allParameters.XY file in your plain text editor</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>Open the new <strong>allParameters.XY</strong> file created above in your plain text editor
for editing. For example, you might open the <em>nano</em> text editor from the command
line:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>nano allParameters.XY</code></pre>
</div></div>
<div class="paragraph"><p>If you are in a hurry, you don&#8217;t need to read anything in the file, but
can simply <strong>search for "##\#"</strong>, which are comment lines
marking items that may need to be changed.</p></div>
<div class="paragraph"><p>Each "#@@" comment says whether it must be changed, might need to
be changed, or probably will never need to be changed, etc. Many of these items
typically will never need to be changed, so the actual number of changes that need
to be made is smaller than it might first appear.  It is
recommended that rather than hurrying, you take time to read through
the file, as the comments explain the purpose of each parameter, and you will
want to know this information, at least for key parameters, to select the right
parameter values for your needs.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Search for "#@@" and set parameters to desired values</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>Search for "#@@" in the allParameters.XY file and check each one to see if it
needs to be changed. If so, set it to the value you desire. Parameters you will
definitely want to review and consider are:</p></div>
<div class="olist loweralpha"><ol class="loweralpha">
<li>
<p>
K
</p>
</li>
<li>
<p>
N_GENOMES
</p>
</li>
<li>
<p>
GENOME_NUMBERS
</p>
</li>
<li>
<p>
GENOME_1, GENOME_2, etc.
</p>
</li>
<li>
<p>
DIR_IGGPIPE_OUT
</p>
</li>
<li>
<p>
PATH_GENOME_FASTA_1, PATH_GENOME_FASTA_2, etc.
</p>
</li>
<li>
<p>
LMIN
</p>
</li>
<li>
<p>
DMAX
</p>
</li>
<li>
<p>
AMIN and AMAX
</p>
</li>
<li>
<p>
ADMIN and ADMAX
</p>
</li>
<li>
<p>
NDAMIN
</p>
</li>
<li>
<p>
OVERLAP_REMOVAL
</p>
</li>
<li>
<p>
EPCR_MAX_DEV
</p>
</li>
<li>
<p>
EPCR_MAX_MISMATCH and EPCR_MAX_GAPS
</p>
</li>
</ol></div>
<div class="paragraph"><p>The next section gives guidance on choosing a value for K.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Choosing a value for K</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The value of K must be chosen carefully. The larger the value, the more common
unique k-mers will be found, up to a point, beyond which the number will fall
because unique k-mers will begin to be long enough to no longer be in common
with the other genome.
The computational demands of IGGPIPE in the steps that immediately follow the
search for common unique k-mers are directly proportional to the number of such
k-mers, which argues for keeping K as small as possible while not so small that
there won&#8217;t be enough common unique k-mers for marker identification. We have
successfully used 7 to 10 million common unique k-mers for producing markers;
several hundred thousand might produce too few markers, while several tens of
millions might create an intolerable computational demand.</p></div>
<div class="paragraph"><p>The optimum value of
K depends on the genetic architecture and amount of polymorphism between
genomes. Very different genomes will have a peak number of common unique k-mers
at a smaller K, so a smaller value would be a good choice. However, this does
not necessarily mean that very similar genomes should use a larger value of K.
Two <em>Arabidopsis thaliana</em> accessions will produce a much larger number of common
unique k-mers for a given K, since most unique k-mers will be common. The challenge
in that case is not to find enough of them, but to avoid having so many that
computational resources are overwhelmed.  We found that a value of 14 worked well
for the quite different <em>S. lycopersicum</em> and <em>S. pennellii</em> genomes, but for the
Arabidopsis accessions we
tested, a value of 14 produced so many common unique k-mers that computation
time was relatively long, while a value of 13 produced almost as many IGG
markers in a much shorter time (see the table in the IGGPIPE paper that shows
computation time for these two values of K in Arabidopsis accessions). Polyploid
species might require a larger value of K to locate enough unique k-mers in the
repeated genomes.</p></div>
<div class="paragraph"><p>We advise checking the number of common unique k-mers obtained with a given value
of K the first time the pipeline is run on a given set of genomes, and adjust K
if necessary. To speed this process, the pipeline can be run only through the
stage where the common unique k-mers are produced, as described below. If too few
k-mers result,
then the user should increase K by one, and if too many k-mers result, the user
should decrease K by one. The pipeline is then run to completion to produce the
IGG marker primer sequences, and the total number of primer pairs produced is
examined.  If there are too few, it is advisable to rerun the entire pipeline
with K both increased and decreased by one.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Additional notes on setting parameter values</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The parameter DIR_IGGPIPE_OUT is the path of the folder where all output will be
placed.  You don&#8217;t need to change this parameter but you should note what its value
will be, and change it if you want to.</p></div>
<div class="paragraph"><p>The parameters PATH_GENOME_FASTA_1, PATH_GENOME_FASTA_2, and so on, up to the number
of genomes being analyzed, are the paths to the FASTA files for the genomes to be
analyzed.  If any of these contain scaffolds, you may want to consider whether you
should remove smaller scaffolds.  The e-PCR portion of the pipeline takes an
inordinately long amount of time to run when there are tens to hundreds of thousands
of scaffolds.  If this is your case, and if many of the scaffolds are very small and
not likely to contribute to viable markers, you should remove them from the FASTA file
to be analyzed.</p></div>
<div class="paragraph"><p>After finishing changes, save the modified allParameters.XY file.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Check Primer3 settings in primer3settings.txt if desired</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The file <strong>primer3settings.txt</strong> contains parameter settings for Primer3, which
is used to generate the actual primers. It is possible that you might want to
use different Primer3 settings from the defaults listed in this file. If so,
edit the file and make the desired changes. For example, you might change the
parameters that determine the acceptable <em>range of primer Tm values</em>. If you
have several different setting values you use, you will probably want to keep
a directory of different primer3settings.txt files and copy the needed one
to "primer3settings.txt" prior to each run of IGGPIPE. When testing IGGPIPE
using the allParameters.test file that was created during installation, always
copy primer3settings.default.txt to primer3settings.txt first.</p></div>
<div class="paragraph"><p>The Primer3 user manual (<a href="http://primer3.sourceforge.net/primer3_manual.htm">http://primer3.sourceforge.net/primer3_manual.htm</a>)
describes all the Primer3 settings file parameters.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Understand use of <em>make</em> and "Makefile" for running IGGPIPE</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The IGGPIPE software consists of multiple software applications that progressively
analyze the genome sequence data and eventually produce candidate IGG marker
primers. The task of running all this software has been automated using a
"Makefile", which is a file with that name containing commands formatted correctly
for reading the allParameters.XY parameter file and running the software applications.
The Makefile is applied by using the application named <em>make</em>, which was installed
when IGGPIPE was installed, if it didn&#8217;t already exist.</p></div>
<div class="paragraph"><p>A big advantage of using "Makefile" and <em>make</em> is that if something goes wrong
(and by Murphy&#8217;s law, it probably will), the portion of the work successfully
completed is not lost, and does not need to be repeated. This is important because
it can take quite a long time to run genomes all the way through the IGGPIPE
software. Depending on your computer speed and memory, it can take hours or even
days. If an error occurs, <em>make</em> will stop, and an error message should be visible
on the terminal. After fixing the error, all you have to do resume the pipeline
commands from the last successful step is re-enter the same <em>make</em> command.
<em>make</em> knows which step to start at because it knows all the files to be produced
by the pipeline, as they are specified in allParameters.XY, and it checks to see
if the files exist, and starts at the pipeline step whose output file does not
exist. Manually deleting a file will also cause <em>make</em> to run the pipeline
starting at the step needed to make that file.</p></div>
<div class="paragraph"><p>You must finish editing the allParameters.XY file before trying to run the pipeline
using <em>make</em>. If that file is ready to go, you can start running IGGPIPE using the
command <em>make</em> from the command line, with additional command arguments. The first
argument that is required is of the form "PARAMS=&lt;allParameters filename&gt;". For
example, if your allParameters file is named "allParameters.XY", then the <em>make</em>
command starts out as <em>make PARAMS=allParameters.XY</em>.</p></div>
<div class="paragraph"><p>The remaining command arguments for the <em>make</em> command tell which part of the
pipeline to run. If no additional argument is given or if the argument is <em>ALL</em>,
the entire pipeline is run (or as
much of it as is needed to resume where a previous error had halted). However,
since the choice of some of the parameters, especially the value of K, can have
a strong influence on the number of markers found, it is best to run IGGPIPE a
few steps at a time and check the output after those steps before proceeding
further. The following sections will guide you in this.</p></div>
<div class="paragraph"><p>Use this command to get a listing of complete usage information for running
<em>make</em>:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make usage</code></pre>
</div></div>
<div class="paragraph"><p>That command will use the <em>less</em> command to display file <em>help.txt</em>. Press the
space bar to move through the text, or press <em>q</em> to exit from the help text.</p></div>
<div class="paragraph"><p>If at any point you want to remove ALL files already generated and start anew,
you can do that with this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY CLEAN=1 ALL    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>Running IGGPIPE with a <em>make</em> command will usually produce a lot of output on
the terminal, and some of this output may be important to examine, especially
if an error occurs. Since the output might scroll off the screen and be
unavailable, it is a good idea to save it, and this can be done by using
the <em>tee</em> command along with the <em>make</em> command. The <em>tee</em> command can write
everything that is displayed on the terminal to a file also. You might want to
make a folder to contain these "log" files:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>mkdir logFiles</code></pre>
</div></div>
<div class="paragraph"><p>To use <em>tee</em>, choose a log file name, let&#8217;s say <em>make_HP14.txt</em>, and then add at
the end of your <em>make</em> command line the extra
commands <em>| tee logFiles/make_HP14.txt</em>, as in this example:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY ALL | tee logFiles/make_HP14.txt</code></pre>
</div></div>
<div class="paragraph"><p>Then, after <em>make</em> finishes, you can examine that log file at any time to see
what the pipeline output was, for example:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>less logFiles/make_HP14.txt</code></pre>
</div></div>
<div class="paragraph"><p>You should use the <em>tee</em> command each time you
run the pipeline unless you are sure you won&#8217;t want to reexamine the output
later. We will not show the <em>tee</em> command in the instructions below, however.
It is up to you to decide whether to use it.</p></div>
<div class="paragraph"><p>We have run IGGPIPE on several different genomes to try to anticipate unusual
problems and handle them without error, but there are probably many situations
that we haven&#8217;t yet encountered. If you email us with information about errors
and their resolution if you were able to resolve them, we&#8217;ll try to make
improvements to IGGPIPE in error handling and in its input data format
flexibility to help future users that encounter the error.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Run IGGPIPE with the command "make PARAMS=allParameters.XY getGenomicPosIsect GENOME=1"</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The first several steps in the pipeline extract unique k-mers from the FASTA
files of the genomes, intersect these to produce a list of common unique
k-mers, and add genomic positions to them for each of the genomes. To run these
steps, use this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>cd ~/Documents/IGGPIPE    (or whatever is appropriate for your system)
make PARAMS=allParameters.XY getGenomicPosIsect GENOME=1    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>or, better yet, log to a file also:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY getGenomicPosIsect GENOME=1 | \
tee logFiles/make_getGenomicPosIsect_HP14.txt</code></pre>
</div></div>
<div class="paragraph"><p>If it completes successfully, the end of the command output will show the message:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>getGenomicPosIsect file(s) for genome(s) 1 are up to date.</code></pre>
</div></div>
<div class="paragraph"><p>If it says something else, indicating an error occurred, examine the output
carefully and try to diagnose and fix the error, then enter the above <em>make</em>
command again to retry the failed step.</p></div>
<div class="paragraph"><p>Assuming the first steps completed successfully, count the number of k-mers in
the common unique k-mer file for genome 1, which is located in the <em>Kmers</em> subfolder
of the output folder you specified in your allParameters file for parameter
DIR_IGGPIPE_OUT.
The name of the file is <em>Kmers_1.isect</em> and it is a text file containing one k-mer
per line, with the position information for that k-mer following the k-mer on the line.
You can look at it with the <em>less</em> command if you want to. To count the number of
k-mers in it, use the <em>wc -l</em> command, which counts lines in a file, like this
(replacing <em>outFolderForMyProject</em> with your output folder name):</p></div>
<div class="literalblock">
<div class="content">
<pre><code>wc -l outFolderForMyProject/Kmers/Kmers_1.isect</code></pre>
</div></div>
<div class="paragraph"><p>If it shows that you have, say, several million or more, that is good. A few
million or less might be too few to generate enough markers. Tens of millions
might be too many and cause subsequent pipeline steps to take a very long time.
The number of k-mers is influenced by both the value of K in the parameter file
and by how different the genomes are. Very similar genomes might never have
several millions of k-mers. The number of common unique k-mers increases as K
is increased, up to a point, then starts to decrease. For similar genomes, you
might need to edit the parameter file and decrease K by 1. With the tomato/pennellii
genomes, K=14 was a
good number. With <em>Arabidopsis thaliana</em> Col-0 and Ler-0 accessions, which were
much more similar to one another than tomato/pennellii, a value of K=13 worked
better, because K=14 produced so many k-mers that computation time was excessive.
For very large genomes, you might need to increase K by 1. Since the
number of k-mers goes up dramatically with increasing K, you probably will never
raise K above 15, perhaps 16. The default setting in the parameter file for the
output directory parameter DIR_IGGPIPE_OUT is to include the value of K in the
directory name. This means you can run IGGPIPE with one value of K, then change
it and run it again and the output will go into a new directory.</p></div>
<div class="paragraph"><p>If you feel you have too few or too many k-mers, then you should increase or
decrease K by 1 and try again. Try both an increase and a decrease in K to see
how the number of k-mers is affected. Set the parameter file to the value of K
that seems best to you for proceeding with additional pipeline steps.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Run IGGPIPE with the command "make PARAMS=allParameters.XY findLCRs"</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The next few steps of the pipeline analyze the common unique k-mers to find
locally conserved regions (LCRs). To run these steps, use this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY findLCRs    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>If it completes successfully, the end of the command output will show the message:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>findLCRs files are up to date.</code></pre>
</div></div>
<div class="paragraph"><p>If it says something else, indicating an error occurred, examine the output
carefully and try to diagnose and fix the error, then enter the above <em>make</em>
command again to retry the failed step.</p></div>
<div class="paragraph"><p>Assuming the steps completed successfully, count the number of LCRs in the LCRs
output file, which is located in the main output directory (set with the DIR_IGGPIPE_OUT
parameter). Its name starts with <em>LCRs_</em>, and with many command line interfaces
you don&#8217;t need to enter the full name in a command, but can instead enter <em>LCRs_*</em>.
The LCRs file has one LCR per line. You can look at the first few lines with the
<em>head</em> command or load the file into Excel or a text editor to examine it, if you
want. The command line to count the LCRs would look like this:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>wc -l outFolderForMyProject/LCRs*</code></pre>
</div></div>
<div class="paragraph"><p>If you had too few common k-mers you might also have too few LCRs.  A million or
more LCRs would be nice. The fewer you have, the fewer markers you are likely to
get.  If there seem to be too few, check the pipeline output.  It will show the
number of common unique k-mers it processes (it processes them in batches), and
the number remaining after it enforces DMIN, LMIN, and KMIN on the reference
genome. If these numbers fall dramatically towards 0, it indicates that either
there are no good LCRs between the two genomes, or the parameters DMIN, LMIN, and/or
KMIN might be too strict. However, expect a pretty big drop with the LMIN step,
because typically a large fraction of the common unique k-mers are too close
together, with too much separation from the next k-mer, to form a useful LCR.</p></div>
<div class="paragraph"><p>If you feel you have too few LCRs, then you should try editing the parameter file
and changing the DMIN, LMIN,
and/or KMIN parameters and try again. The default value for the LCRs_ filename,
set by the parameters SFX_LCR_FILE and PATH_LCR_FILE, includes the values of
DMIN, LMIN, and KMIN in the filename, so if you change the values, you can
simply re-run the pipeline with the same <em>make</em> command, and it will generate a
new LCRs_ file with a different name, without repeating preceding pipeline steps
that do not need to be repeated.</p></div>
<div class="paragraph"><p>Set the parameter file to the values for DMIN, LMIN, and KMIN that seem best to
you and re-run this pipeline step before proceeding with additional pipeline
steps.</p></div>
<div class="paragraph"><p>The columns in the LCRs* file are described in Table 4.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Run IGGPIPE with the command "make PARAMS=allParameters.XY findIndelGroups"</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The next step of the pipeline analyzes the LCRs to find Indel Groups that satisfy
the parameters AMIN, AMAX, ADMIN, ADMAX, NDAMIN, and MINFLANK. To run this step,
use this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY findIndelGroups    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>If it completes successfully, the end of the command output will show the message:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>findIndelGroups files are up to date.</code></pre>
</div></div>
<div class="paragraph"><p>If it says something else, indicating an error occurred, then as usual, examine
the output carefully and try to diagnose and fix the error, then enter the
above <em>make</em> command again to retry the failed step.</p></div>
<div class="paragraph"><p>Assuming the step completed successfully, count the number of Indel Groups in the
two output files. One output file includes all Indel Groups found, even when they
overlap one another. The other output file includes only non-overlapping Indel
Groups, which were determined based on the setting of the parameter OVERLAP_REMOVAL.
The output files are located in the main output directory and their names start
with <em>IndelGroups</em> followed by <em>Overlapping_</em> and <em>Nonoverlapping_</em>.
You can look at the first few lines with the <em>head</em> command or
load the files into Excel or a text editor to examine them, if you want.
The command line to count the Indel Groups in both files would look like this:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>wc -l outFolderForMyProject/IndelGroups*</code></pre>
</div></div>
<div class="paragraph"><p>The number of Indel Groups is of the same order of magnitude as the number of
markers you will obtain, so pay close attention to the count. If you had too
few LCRs you might also have too few markers. Tens of thousands of overlapping
markers and thousands of non-overlapping ones are nice numbers to have. If
there seem to be too few, you may want to experiment with different values for
the AMIN, AMAX, ADMIN, ADMAX, NDAMIN, and MINFLANK parameters. As with the LCRs,
these values are normally part of the IndelGroups output file filenames, so if
you edit the parameter file and change the values, then re-run the pipeline by
entering the same <em>make</em> command above, new files are produced with new names.</p></div>
<div class="paragraph"><p>Set the parameter file to the values for the parameters that seem best to
you and re- run this pipeline step before proceeding with remaining pipeline steps.</p></div>
<div class="paragraph"><p>The columns in the IndelGroups* files are described in Table 3.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Run IGGPIPE with the command "make PARAMS=allParameters.XY findPrimers"</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The next several steps of the pipeline extract DNA sequences for each Indel Group
from all genomes using the parameter EXTENSION_LEN, then Primer3 (actually, primer3_core)
is run to design primers for each Indel Group, using the primer design parameters
in file primer3settings.txt.</p></div>
<div class="paragraph"><p>To run these steps, use this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY findPrimers    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>Depending on the number of Indel Groups, this can take a <em>long</em> time. We have seen it
take two days to finish running all Indel Groups through Primer3.</p></div>
<div class="paragraph"><p>If it completes successfully, the end of the command output will show the message:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>findPrimers files are up to date.</code></pre>
</div></div>
<div class="paragraph"><p>If it says something else, try to diagnose and fix the error as usual, then enter
the same <em>make</em> command again to retry the failed step.</p></div>
<div class="paragraph"><p>Assuming the steps completed successfully, count the number of IGG marker primer
pairs in the output file, which is located in the main output directory and has a
name that starts with <em>NonvalidatedMarkers_</em>.
You can look at the first few lines with the <em>head</em> command or
load the files into Excel or a text editor to examine them, if you want.
The command line to count the primer pairs in the file would look like this:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>wc -l outFolderForMyProject/NonvalidatedMarkers*</code></pre>
</div></div>
<div class="paragraph"><p>Each line of the file contains one pair of primers.  Each pair is a candidate
IGG marker, but they have not yet been validated using e-PCR, which will mark a
few of them as bad and remove them.</p></div>
<div class="paragraph"><p>The columns in the NonvalidatedMarkers* file are described in Table 1.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Run IGGPIPE with the command "make PARAMS=allParameters.XY removeBadMarkers"</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The next several steps of the pipeline run e-PCR on each primer pair in the
NonvalidatedMarkers* file.  All markers are run through e-PCR once for each genome.
After that, an R script is run which examines the e-PCR results and removes from
the NonvalidatedMarkers* primer pairs all those pairs that failed the e-PCR test
in or more genomes, and writes new files with the validated IGG marker primer pairs.</p></div>
<div class="paragraph"><p>To run these steps, use this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY removeBadMarkers    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>Depending on the number of primer pairs, this can take a <em>long</em> time. We have seen it
take two days <em>per genome</em> to finish running all primer pairs through e-PCR.</p></div>
<div class="paragraph"><p>If it completes successfully, the end of the command output will show the message:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>removeBadMarkers files are up to date.</code></pre>
</div></div>
<div class="paragraph"><p>If it says something else, try to diagnose and fix the error as usual, then enter
the same <em>make</em> command again to retry the failed step.</p></div>
<div class="paragraph"><p>Assuming the steps completed successfully, count the number of validate IGG primer
pairs in the two output files. One output file includes all validated primer pairs,
even when their amplicons overlap one another. The other output file includes only
primer pairs that produce non-overlapping amplicons, determined based on the setting
of the parameter OVERLAP_REMOVAL.
The output files are located in the main output directory and their names start
with <em>Markers</em> followed by <em>Overlapping_</em> and <em>Nonoverlapping_</em>.
You can look at the first few lines with the <em>head</em> command or
load the files into Excel or a text editor to examine them, if you want.
The command line to count the primer pairs in both files would look like this:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>wc -l outFolderForMyProject/Markers*</code></pre>
</div></div>
<div class="paragraph"><p>Each line of the file contains one pair of primers.  Each pair is an e-PCR-validated
IGG marker, essentially the final output of the pipeline.</p></div>
<div class="paragraph"><p>The columns in the Markers* file are described in Table 1.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Run IGGPIPE with the command "make PARAMS=allParameters.XY ALL"</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>Now you can run the rest of the pipeline to finish up.  The final step produces some
plots of marker statistics and density.  This is done with this command:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY plotMarkers    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>However, you can also use this command, which is the command that runs the entire
pipeline, all steps from start to finish:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY ALL    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>This same command could have been used from the very start, to simply run the
entire pipeline. It can be used <em>at any time</em> to re-run the pipeline.  For each
step of the pipeline that was previously run successfully, it will simply output
a message saying that the files from that pipeline step are up to date.  However,
for any step that was either not run at all, or failed, it will attempt to re-run
that step and following steps that depend on it.  At the very end, when it has
successfully completed all pipeline steps, it issues this message:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>ALL files are up to date</code></pre>
</div></div>
<div class="paragraph"><p>This is the message you want to see to know that the pipeline has completed
successfully.</p></div>
<div class="paragraph"><p>If you edit the parameter file and change parameters (so that the names of the
files that are produced are also changed, since the parameter values are contained
within the file names), then when you use the above <em>make</em> command, the pipeline
re-runs starting at whatever step uses the changed parameters, so you can easily
make parameter changes and try again.</p></div>
<div class="paragraph"><p>Sometimes you may want to force the pipeline to re-run starting at a certain
step. For example, maybe you want to re-run the pipeline starting at the <em>findLCRs</em>
step. To do this, you can either delete the output files produced by that step,
or use this command to delete them:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY CLEAN=1 findLCRs    (replacing 'findLCRs' with whatever step you want)</code></pre>
</div></div>
<div class="paragraph"><p>After that, you can re-run the pipeline with the <em>make &#8230; ALL</em> command shown above
to generate the files anew.</p></div>
<div class="paragraph"><p>The final step, plotMarkers, produces several output files in the main output directory.
One file is a .pdf file whose name starts with MarkerCounts_, containing plots of
counts of markers on each chromosome of each genome.  You can examine that file with
any .pdf file viewer.  The other files are .png image files whose names start with
MarkerDensity_ and end with <em>_X.plot.png</em>, where X is replaced with the genome letters
you assigned for your analysis.  Each .png file has an image of the chromosomes with
lines showing the positions of each marker.  When multiple scaffolds are used, the
file limits the output to the first several scaffolds.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Comments about pipeline result files</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>Here is a summary of the filenames produced by each <em>make</em> step, in pipeline order:</p></div>
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<col width="50%" />
<col width="50%" />
<thead>
<tr>
<th align="left" valign="top"><em>make</em> command or other command</th>
<th align="left" valign="top">Produces output file(s)</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left" valign="top"><p class="table">a. make PARAMS=myFilename getSeqInfo</p></td>
<td align="left" valign="top"><p class="table">GenomeData/*.idlens</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">b. make PARAMS=myFilename getContigFile</p></td>
<td align="left" valign="top"><p class="table">GenomeData/*.contigs</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">c. make PARAMS=myFilename getKmers</p></td>
<td align="left" valign="top"><p class="table">Kmers/Kmers_*.kmers</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">d. make PARAMS=myFilename kmerStats</p></td>
<td align="left" valign="top"><p class="table">Kmers/Kmers_*.stats</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">e. make PARAMS=myFilename kmersToText</p></td>
<td align="left" valign="top"><p class="table">Kmers/Kmers_*.kmers.txt</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">f. make PARAMS=myFilename getGenomicPosIsect</p></td>
<td align="left" valign="top"><p class="table">Kmers/Kmers_*.isect</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">g. make PARAMS=myFilename mergeKmers</p></td>
<td align="left" valign="top"><p class="table">Kmers/Kmers_*.merge</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">h. make PARAMS=myFilename sortCommonUniqueKmers</p></td>
<td align="left" valign="top"><p class="table">Kmers/common.unique.kmers</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">i. make PARAMS=myFilename findLCRs</p></td>
<td align="left" valign="top"><p class="table">LCRs_*.tsv, BadKmers_*.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">j. make PARAMS=myFilename findIndelGroups</p></td>
<td align="left" valign="top"><p class="table">IndelGroupsOverlapping_*.tsv, IndelGroupsNonoverlapping_*.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">k. make PARAMS=myFilename getDNAseqsForPrimers</p></td>
<td align="left" valign="top"><p class="table">IndelGroupsOverlapping_*.dnaseqs</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">l. make PARAMS=myFilename findPrimers</p></td>
<td align="left" valign="top"><p class="table">NonvalidatedMarkers_*.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">m. make PARAMS=myFilename ePCRtesting</p></td>
<td align="left" valign="top"><p class="table">MarkerErrors_*.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">n. make PARAMS=myFilename removeBadMarkers</p></td>
<td align="left" valign="top"><p class="table">MarkersOverlapping_*.tsv, MarkersNonoverlapping_*.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">o. make PARAMS=myFilename plotMarkers</p></td>
<td align="left" valign="top"><p class="table">MarkerCounts_*.pdf, MarkerDensity_*.png</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">p. make PARAMS=myFilename getDNAseqsForIndelsSNPs</p></td>
<td align="left" valign="top"><p class="table">*.withseqs.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">q. make PARAMS=myFilename IndelsSNPs</p></td>
<td align="left" valign="top"><p class="table">*.indels.tsv, *.snps.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">r. make PARAMS=myFilename plotIndels</p></td>
<td align="left" valign="top"><p class="table">*.indels.pdf</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">s. Rscript code/R/dotplot.R dotplot.template</p></td>
<td align="left" valign="top"><p class="table">LCRs_*.dotplot.png</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">t. Rscript code/R/annotateFile.R annotate.template</p></td>
<td align="left" valign="top"><p class="table">MarkersAnnotated_*.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">u. Rscript code/R/annotateFile.R annotate/HP11_isInNearColumn.markers</p></td>
<td align="left" valign="top"><p class="table">MarkersAnnotated_WithInNearFeatures_*.indels.tsv</p></td>
</tr>
<tr>
<td align="left" valign="top"><p class="table">v. Rscript code/R/annotateFile.R annotate/HP11_to_gff3.markers</p></td>
<td align="left" valign="top"><p class="table">MarkersAnnotated_GFF3_*.gff3</p></td>
</tr>
</tbody>
</table>
</div>
<div class="paragraph"><p>(Note that some of the files listed above are produced by steps to be described below).</p></div>
<div class="paragraph"><p>The marker file names and other file names of files in the main output directory
are very long and cumbersome, because they include parameter values in them. You
may want to copy files to a shorter name to work with them.</p></div>
<div class="paragraph"><p>The meaning of <em>overlapping</em> and <em>non-overlapping</em> should be clear from the
explanation of the parameter OVERLAP_REMOVAL in the comments in allParameters.XY.</p></div>
<div class="paragraph"><p>The various .tsv files can be loaded into Excel to examine, and they can also be
post-processed (see below) to change them into other formats.</p></div>
<div class="paragraph"><p>Tables describing each column in each .tsv file type are at the end of this document.</p></div>
</div></div>
</li>
</ol></div>
</div>
</div>
<div class="sect1">
<h2 id="_strong_post_processing_tools_strong"><strong>Post-processing tools</strong></h2>
<div class="sectionbody">
<div class="olist arabic"><ol class="arabic">
<li>
<p>
<strong>Finding Indels and SNPs</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>Pipeline software is also provided to read a file of
LCRs, non-overlapping Indel Groups, or non-overlapping Markers, extract the DNA sequences
from the genomes in each LCR or Marker region and align them, then locate all Indels
and SNPs in the aligned sequences and write their positions to files.  This part of the
pipeline is NOT run when the <em>make &#8230; ALL</em> target is built.  To run this and find
Indels and SNPs, use <em>make &#8230; IndelsSNPs</em>, after setting the parameters in your
"allParameters" parameter file.  These are PATH_INDELS_SNPS_INPUT_FILE (the input file
name), MAX_INDELS_PER_KBP and MAX_SNPS_PER_KBP (maximum number of Indels and SNPs that
may occur in an alignment per Kbp of sequence, and if exceeded, cause the alignment to
be discarded as unreliable), and SCRAMBLE_SEQUENCE (can be set TRUE to do alignments
with scrambled sequences, to determine the actual numbers of Indels and SNPs per Kbp
in random sequence alignments.  Run the Indel/SNP finder as follows:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY IndelsSNPs    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>This produces two files in your output folder whose names end with "indels.tsv" and
".snps.tsv", containing tables of all Indels and SNPs found in either the LCR or Indel
Group or marker regions (depending on the setting of PATH_INDELS_SNPS_INPUT_FILE).
Examine them to see the data they contain.  The columns are described in Table 5 and
Table 6.</p></div>
<div class="paragraph"><p>The default settings for the parameters MAX_INDELS_PER_KBP and MAX_SNPS_PER_KBP are
reasonable values, but you can adjust them to either decrease FALSE positives at the
expense of fewer TRUE positives (smaller values) or increase TRUE positives at the
expense of more FALSE positives (larger values).</p></div>
<div class="paragraph"><p>Note that the pipeline for <em>make IndelsSNPs</em> automatically invokes a preceding step,
<em>make getDNAseqsForIndelsSNPs</em> that extracts DNA sequences in preparation for alignment,
and writes them to a file whose name ends with "withseqs.tsv".</p></div>
</div></div>
</li>
<li>
<p>
<strong>Plotting Indel information</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>Another R program that is NOT run as part of the pipeline when the <em>make &#8230; ALL</em> target
is built, but which can be run using <em>make &#8230; plotIndels</em>, reads the Indels file produced
by <em>make &#8230; IndelsSNPs</em> and plots information from it in a pdf file. The program is called
plotIndels.R. Run it as follows:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>make PARAMS=allParameters.XY plotIndels    (replacing with your allParameters name)</code></pre>
</div></div>
<div class="paragraph"><p>This produces a file in your output folder whose name ends with "indels.pdf",
containing plots of various Indel information.  Examine it to see the plots it
contains.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Dot plots</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>The LCRs_*.tsv output file contains
locally conserved regions associated with common unique k-mers. It represents a
whole genome alignment between the genomes used in IGGPIPE analysis. An R program,
dotplot.R, is provided that can plot this data as a dot plot.</p></div>
<div class="paragraph"><p>This program is run
by first copying the text file "dotplot.template" to a new name (e.g. dotplot.XY)
and editing it to specify the parameters of the dot plot. Comments in the file
describe each parameter.  The program is then run from
the command line with a command like this:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>cd ~/Documents/IGGPIPE    (or whatever is appropriate for your system)
Rscript code/R/dotplot.R dotplot.XY    (or whatever name you gave the parameter file)</code></pre>
</div></div>
<div class="paragraph"><p>When it finishes running, the dot plot output file can be found in the place and under
the name specified in the parameter file. Use multiple parameter files with different
settings to explore different regions of the genomes in greater resolution (parameters
include what region of the genome is to be plotted).</p></div>
<div class="paragraph"><p>The "dotplot.template" file is configured for generating a dot plot file
using the LCRs generated via the allParameters.test configuration file.</p></div>
<div class="paragraph"><p>Besides the sample parameter file "dotplot.template" (which has settings for
testing the IGGPIPE installation), there are several more sample "dotplot" parameter
files in the subfolder <em>dotplot</em>, that were used for plotting data from various
genomes that IGGPIPE was tested with, although the
parameter file is straightforward and you probably don&#8217;t need other examples
to work from.  You may want to put your own "dotplot"
parameter files in subfolder <em>dotplot</em> or your own subfolder to keep them organized.</p></div>
</div></div>
</li>
<li>
<p>
<strong>Annotating marker files with other position data and producing GFF3 and GTF files</strong>
</p>
<div class="openblock">
<div class="content">
<div class="paragraph"><p>You may want to make your marker data more conveniently available.  For example,
you might want to convert it to GFF3 file format so you can add a "marker" track
to a genome browser.  Or, you may have other genome position data that you would
like to have associated with your marker data, such as a file giving positions
of introgressions of one genome within another (you might want a column in the
marker file showing which introgressions the marker was near). As another example,
you might want to add a column in the marker file containing the names of the
genes closest to the marker, and the distance to the genes.  All of these
situations and more can be handled by an R program, annotateMarkers.R, provided
with IGGPIPE. The program can read and write files of type .tsv (tab-separated
variable), .csv (comma-separated variable), .gff3 (general feature format), or
.gtf (gene transfer format), all common formats used to hold genome browser track
data or FASTA file annotation data.  It can add, remove, edit, and rename columns.
It can read two separate files and merge their data.  It can convert from one of
these file formats to another.</p></div>
<div class="paragraph"><p>This program is run by first copying the text file "annotate.template" to a new
name (e.g. annotateIntrogressions.XY or addGeneInfo.XY or makeGFF3.XY) and then
editing it to specify the parameters for the annotation and/or file conversion.
Comments in the file describe each parameter. The program is then run from the
command line with a command like this:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>cd ~/Documents/IGGPIPE    (or whatever is appropriate for your system)
Rscript code/R/annotate.R addGenes.XY    (or whatever name you gave the parameter file)</code></pre>
</div></div>
<div class="paragraph"><p>When it finishes running, the output files can be found in the place(s) and under
the name(s) specified in the parameter file.</p></div>
<div class="paragraph"><p>Besides the sample parameter file "annotate.template" (which has settings for
testing the IGGPIPE installation), there are several more sample "annotate"
parameter files in
the subfolder <em>annotate</em>, with file names hinting at what they do, and comments at
the start of each file describing what it does.  It may be easier to copy one of
these and modify it for your needs.  You may want to put your own "annotate"
parameter files in subfolder <em>annotate</em> or your own subfolder to keep them organized.</p></div>
<div class="paragraph"><p>So, the idea is to use multiple parameter files with different settings to do
different types of annotation and file conversion.</p></div>
<div class="paragraph"><p>Some of the sample parameter files generate .gff3 files that can be added as a
track to a genome browser, to display markers in the browser.  Instructions for
adding the track are given in comments at the start of the parameter file.  Two
marker files, one for <em>Arabidopsis thaliana</em> Col-0 vs. Ler-0 accessions, and the
other for <em>Solanum lycopersicum</em> vs. <em>Solanum pennellii</em> genomes, were created to
test IGGPIPE, and the marker files were converted to .gff3 files suitable for
making a browser track.  These files can be found in subfolders of the
<em>annotate</em> folder.</p></div>
<div class="paragraph"><p>File formats can be finicky, especially .gff3 files.  An incorrectly formatted
file will cause problems with annotateFile.R.  When you have problems, if you
can submit an issue to the GitHub repository named "BradyLab/IGGPIPE", and attach
or insert a copy of your parameter file, that would be helpful.  A copy of the
input data files would probably also be needed to debug problems, but GitHub
does not allow files to be attached.  You can email them to us, or find some
other way to send them.</p></div>
</div></div>
</li>
</ol></div>
<div class="sect2">
<h3 id="_strong_for_problems_and_help_strong"><strong>For problems and help:</strong></h3>
<div class="ulist"><ul>
<li>
<p>
Post an issue on GitHub under BradyLab/IGGPIPE repository
</p>
</li>
<li>
<p>
Contact: Ted Toal, <a href="mailto:twtoal@ucdavis.edu">twtoal@ucdavis.edu</a>
</p>
</li>
</ul></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_strong_tables_strong"><strong>Tables</strong></h2>
<div class="sectionbody">
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<caption class="title">Table 1. Columns in MarkersOverlapping_, MarkersNonoverlapping_, NonvalidatedMarkers_, MarkerErrors_ files; X,Y=chosen genome letters</caption>
<col width="10%" />
<col width="90%" />
<thead>
<tr>
<th align="center" valign="top">Column</th>
<th align="left" valign="top">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center" valign="top"><p class="table">NDA</p></td>
<td align="left" valign="top"><p class="table">Number of distinct amplicon sizes, in range NDAMIN..N_GENOMES</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xid</p></td>
<td align="left" valign="top"><p class="table">Genome X sequence ID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xpct</p></td>
<td align="left" valign="top"><p class="table">Genome X percent of sequence ID length at which marker is located</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">XampLen</p></td>
<td align="left" valign="top"><p class="table">Genome X amplicon length</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Yid</p></td>
<td align="left" valign="top"><p class="table">Genome Y sequence ID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Ypct</p></td>
<td align="left" valign="top"><p class="table">Genome Y percent of sequence ID length at which marker is located</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">YampLen</p></td>
<td align="left" valign="top"><p class="table">Genome Y amplicon length</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">YXdif</p></td>
<td align="left" valign="top"><p class="table">Difference in length between genomes X and Y amplicons, negative if genome X longer than genome Y</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">YXphase</p></td>
<td align="left" valign="top"><p class="table">Phase of amplicons between genomes X and Y, "+" if both amplicons run in same direction, "-"
if opposite directions</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">prmSeqL</p></td>
<td align="left" valign="top"><p class="table">Left side or upstream primer sequence</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">prmSeqR</p></td>
<td align="left" valign="top"><p class="table">Right side or downstream primer sequence</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">prmTmL</p></td>
<td align="left" valign="top"><p class="table">Left side primer Tm</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">prmTmR</p></td>
<td align="left" valign="top"><p class="table">Right side primer Tm</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">prmLenL</p></td>
<td align="left" valign="top"><p class="table">Left side primer length</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">prmLenR</p></td>
<td align="left" valign="top"><p class="table">Right side primer length</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">XampPos1</p></td>
<td align="left" valign="top"><p class="table">Genome X amplicon starting (upstream) position</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">XampPos2</p></td>
<td align="left" valign="top"><p class="table">Genome X amplicon ending (downstream) position, XampPos2 always &gt; XampPos1</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">YampPos1</p></td>
<td align="left" valign="top"><p class="table">Genome Y amplicon starting (upstream) position</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">YampPos2</p></td>
<td align="left" valign="top"><p class="table">Genome Y amplicon ending (downstream) position, YampPos2 &gt; YampPos1 if YXphase is "+", &lt; if "-"</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">kmer1</p></td>
<td align="left" valign="top"><p class="table">Common unique k-mer for left side primer region, canonical (exically smaller of k-mer and its reverse complement)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">kmer1strands</p></td>
<td align="left" valign="top"><p class="table">N_GENOMES "<code>" and "-" characters for genomes 1..N_GENOMES. A "</code>" means k-mer 1 lies
on the "+" strand in that genome, "-" means "-" strand.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">kmer1offset</p></td>
<td align="left" valign="top"><p class="table">Offset in bp of outside (away from amplicon) edge of k-mer 1 from that end of the amplicon.
A value of 0 means the amplicon and k-mer ends correspond, &gt;0 means k-mer starts inside the amplicon,
&lt;0 means k-mers starts outside it.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">kmer2</p></td>
<td align="left" valign="top"><p class="table">Common unique k-mer for right side primer region, canonical (exically smaller of k-mer and its reverse complement)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">kmer2strands</p></td>
<td align="left" valign="top"><p class="table">Like kmer1strands, for k-mer 2.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">kmer2offset</p></td>
<td align="left" valign="top"><p class="table">Like kmer1offset, for k-mer 2.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xseq1</p></td>
<td align="left" valign="top"><p class="table">Genome X DNA sequence around left side primer region</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xseq2</p></td>
<td align="left" valign="top"><p class="table">Genome X DNA sequence around right side primer region</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Yseq1</p></td>
<td align="left" valign="top"><p class="table">Genome Y DNA sequence around left side primer region</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Yseq2</p></td>
<td align="left" valign="top"><p class="table">Genome Y DNA sequence around right side primer region</p></td>
</tr>
</tbody>
</table>
</div>
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<caption class="title">Table 2. Column reasonDiscarded in MarkerErrors_ files (see Table 1 for other columns)</caption>
<col width="16%" />
<col width="83%" />
<thead>
<tr>
<th align="center" valign="top">reasonDiscarded</th>
<th align="left" valign="top">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center" valign="top"><p class="table">found multiple</p></td>
<td align="left" valign="top"><p class="table">ePCR found multiple amplicons (expected reason)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">not found</p></td>
<td align="left" valign="top"><p class="table">ePCR didn&#8217;t find amplicon (should never happen)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">wrong seq id</p></td>
<td align="left" valign="top"><p class="table">ePCR sequence ID output is wrong (should never happen)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">wrong pos</p></td>
<td align="left" valign="top"><p class="table">ePCR left and right position output is wrong (should never happen)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">wrong posL</p></td>
<td align="left" valign="top"><p class="table">ePCR left position output is wrong (should never happen)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">wrong posR</p></td>
<td align="left" valign="top"><p class="table">ePCR right position output is wrong (should never happen)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<caption class="title">Table 3. Columns in IndelGroupsOverlapping_ and IndelGroupsNonoverlapping_ files; X,Y=chosen genome letters</caption>
<col width="10%" />
<col width="90%" />
<thead>
<tr>
<th align="center" valign="top">Column</th>
<th align="left" valign="top">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center" valign="top"><p class="table">kmer1</p></td>
<td align="left" valign="top"><p class="table">Common unique k-mer for left side primer region, canonical (lexically smaller of k-mer and its reverse complement)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">kmer2</p></td>
<td align="left" valign="top"><p class="table">Common unique k-mer for right side primer region, canonical (exically smaller of k-mer and its reverse complement)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">NDA</p></td>
<td align="left" valign="top"><p class="table">Number of distinct amplicon sizes, in range NDAMIN..N_GENOMES</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xid</p></td>
<td align="left" valign="top"><p class="table">Genome X sequence ID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xpos1</p></td>
<td align="left" valign="top"><p class="table">Genome X position of upstream end of k-mer 1 on "+" strand</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xpos2</p></td>
<td align="left" valign="top"><p class="table">Genome X position of upstream end of k-mer 2 on "+" strand, Xpos1 &lt; Xpos2 always</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xs1</p></td>
<td align="left" valign="top"><p class="table">Genome X k-mer 1 strand, "+" or "-"</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xs2</p></td>
<td align="left" valign="top"><p class="table">Genome X k-mer 2 strand, "+" or "-"</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xctg1</p></td>
<td align="left" valign="top"><p class="table">Genome X contig number within sequence Xid of contig containing k-mer 1</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xctg2</p></td>
<td align="left" valign="top"><p class="table">Likewise for k-mer 2, Xctg1 = Xctg2 always</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">XkkLen</p></td>
<td align="left" valign="top"><p class="table">Genome X distance from 5' end of k-mer 1 on "<code>" strand to 5' end of k-mer 1 on "</code>" strand</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xpct</p></td>
<td align="left" valign="top"><p class="table">Genome X percent of sequence ID length at which marker is located</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Yid</p></td>
<td align="left" valign="top"><p class="table">Genome Y sequence ID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Ypos1</p></td>
<td align="left" valign="top"><p class="table">Genome Y position of upstream end of k-mer 1 on "+" strand</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Ypos2</p></td>
<td align="left" valign="top"><p class="table">Genome Y position of upstream end of k-mer 2 on "+" strand, Ypos1 &lt; Ypos2
if amplicon in X and Y genomes run in the same direction, &gt; if opposite directions</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Ys1</p></td>
<td align="left" valign="top"><p class="table">Genome Y k-mer 1 strand, "+" or "-"</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Ys2</p></td>
<td align="left" valign="top"><p class="table">Genome Y k-mer 2 strand, "+" or "-"</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Yctg1</p></td>
<td align="left" valign="top"><p class="table">Genome Y contig number within sequence Yid of contig containing k-mer 1</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Yctg2</p></td>
<td align="left" valign="top"><p class="table">Likewise for k-mer 2, Yctg1 = Yctg2 always</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">YkkLen</p></td>
<td align="left" valign="top"><p class="table">Genome Y distance from 5' end of k-mer 1 on "<code>" strand to 5' end of k-mer 1 on "</code>" strand</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Ypct</p></td>
<td align="left" valign="top"><p class="table">Genome Y percent of sequence ID length at which marker is located</p></td>
</tr>
</tbody>
</table>
</div>
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<caption class="title">Table 4. Columns in LCRs_ and BadKmers_ files; X,Y=chosen genome letters</caption>
<col width="10%" />
<col width="90%" />
<thead>
<tr>
<th align="center" valign="top">Column</th>
<th align="left" valign="top">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center" valign="top"><p class="table">(none, row name)</p></td>
<td align="left" valign="top"><p class="table">Common unique k-mer, canonical representation (the lexically smaller of k-mer and its reverse complement)</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">X.seqID</p></td>
<td align="left" valign="top"><p class="table">Genome X sequence ID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">X.pos</p></td>
<td align="left" valign="top"><p class="table">Genome X position of upstream end of k-mer on "+" strand relative to start of X.seqID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">X.strand</p></td>
<td align="left" valign="top"><p class="table">Genome X k-mer strand, "+" or "-"</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">X.contig</p></td>
<td align="left" valign="top"><p class="table">Genome X contig number within sequence X.seqID sequence of contig containing the k-mer</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">X.contigPos</p></td>
<td align="left" valign="top"><p class="table">Genome X position of upstream end of k-mer on "+" strand relative to start of X.contig</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Y.seqID</p></td>
<td align="left" valign="top"><p class="table">Genome Y sequence ID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Y.pos</p></td>
<td align="left" valign="top"><p class="table">Genome Y position of upstream end of k-mer on "+" strand relative to start of Y.seqID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Y.strand</p></td>
<td align="left" valign="top"><p class="table">Genome Y k-mer strand, "+" or "-"</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Y.contig</p></td>
<td align="left" valign="top"><p class="table">Genome Y contig number within sequence X.seqID sequence of contig containing the k-mer</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Y.contigPos</p></td>
<td align="left" valign="top"><p class="table">Genome Y position of upstream end of k-mer on "+" strand relative to start of Y.contig</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">LCR</p></td>
<td align="left" valign="top"><p class="table">Integer LCR number to which this k-mer is assigned, each LCR has a unique LCR number assigned to it</p></td>
</tr>
</tbody>
</table>
</div>
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<caption class="title">Table 5. Columns in *.indels.tsv files; X,Y=chosen genome letters</caption>
<col width="10%" />
<col width="90%" />
<thead>
<tr>
<th align="center" valign="top">Column</th>
<th align="left" valign="top">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center" valign="top"><p class="table">ID</p></td>
<td align="left" valign="top"><p class="table">Unique ID tying row back to originating input file row. LCR input files: LCRnumber.
IndelGroup and Markers files: refID_refPos1_refPos2.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">phases</p></td>
<td align="left" valign="top"><p class="table">Phase of each genome incl. ref. genome, relative to ref. genome, string of +/- chars,
+ : same direction, - : opposite direction.",</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">idx</p></td>
<td align="left" valign="top"><p class="table">Starts at 1 and counts each Indel within an ID. For given ID (input row), number of Indels
in that region is max idx value. If more than two genomes, entire region where alignment has a
gap in one or more genomes is counted as one Indel even if multiple gap regions occur in different
genomes.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xdel,Ydel</p></td>
<td align="left" valign="top"><p class="table">Total number of deleted bps within the Indel in genomes X,Y. With 2 genomes, del = 0
in genome with insertion (no gaps), del &gt; 0 in genome with deletion (gaps). With &gt;2 genomes, del
can be non-zero for all genomes. A genome has only insertions in the Indel if del is 0, and it has
only deletions if end-start-1 = 0, and otherwise it has a mixture of at least one insertion and
one deletion within the Indel interval.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xid,Yid</p></td>
<td align="left" valign="top"><p class="table">Sequence ID of the Indel in genomes X,Y.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xstart,Xend,Ystart,Yend</p></td>
<td align="left" valign="top"><p class="table">Overall Indel starting and ending position in genomes X,Y.
start/end are positions of bps just BEFORE first and AFTER last Indel gap in any genome,
so they refer to the same two bps in all genomes. Always start &lt; end. If <em>-</em> phase,
start is bp just AFTER, end is bp just BEFORE, opposite of <em>+</em>. Length of the Indel region
in each genome is end-start-1.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<caption class="title">Table 6. Columns in *.snps.tsv files; X,Y=chosen genome letters</caption>
<col width="10%" />
<col width="90%" />
<thead>
<tr>
<th align="center" valign="top">Column</th>
<th align="left" valign="top">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center" valign="top"><p class="table">ID</p></td>
<td align="left" valign="top"><p class="table">Unique ID tying row back to originating input file row. LCR input files: LCRnumber.
IndelGroup and Markers files: refID_refPos1_refPos2.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">phases</p></td>
<td align="left" valign="top"><p class="table">Phase of each genome incl. ref. genome, relative to ref. genome, string of +/- chars,
+ : same direction, - : opposite direction.",</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">idx</p></td>
<td align="left" valign="top"><p class="table">Starts at 1 and counts each SNP within an ID. For given ID (input row), number of SNPs
in that region is max idx value.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xid,Yid</p></td>
<td align="left" valign="top"><p class="table">Sequence ID of the SNP in genomes X,Y.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xpos,Ypos</p></td>
<td align="left" valign="top"><p class="table">SNP position in genomes X,Y.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xval,Yval</p></td>
<td align="left" valign="top"><p class="table">SNP value in genomes X,Y.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="tableblock">
<table rules="all"
width="100%"
frame="border"
cellspacing="0" cellpadding="4">
<caption class="title">Table 7. Columns in *.withseqs.tsv files; X,Y=chosen genome letters</caption>
<col width="10%" />
<col width="90%" />
<thead>
<tr>
<th align="center" valign="top">Column</th>
<th align="left" valign="top">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center" valign="top"><p class="table">ID</p></td>
<td align="left" valign="top"><p class="table">Unique ID tying row back to originating input file row. LCR input files: LCRnumber.
IndelGroup and Markers files: refID_refPos1_refPos2.</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">phases</p></td>
<td align="left" valign="top"><p class="table">Phase of each genome incl. ref. genome, relative to ref. genome, string of +/- chars,
+ : same direction, - : opposite direction.",</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xid</p></td>
<td align="left" valign="top"><p class="table">Genome X sequence ID</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xpos1</p></td>
<td align="left" valign="top"><p class="table">Genome X position of upstream end of sequence to align on "+" strand</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xpos2</p></td>
<td align="left" valign="top"><p class="table">Genome X position of downstream end of sequence to align on "+" strand, Xpos1 &lt; Xpos2 always</p></td>
</tr>
<tr>
<td align="center" valign="top"><p class="table">Xseq</p></td>
<td align="left" valign="top"><p class="table">Genome X DNA sequence between the two positions</p></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="footnotes"><hr /></div>
<div id="footer">
<div id="footer-text">
Last updated 2016-01-17 12:31:28 PST
</div>
</div>
</body>
</html>
